pkg/hive/job/job_test.go

```diff
@@ -0,0 +1,757 @@
+	stop := make(chan struct{})
+
+	h := fixture(func(r Registry, l hive.Lifecycle) {
+		g := r.NewGroup()
+
+		g.Add(
+			OneShot("short", func(ctx context.Context) error {
+				defer close(stop)
+				return nil
+			}),
+		)
+
+		l.Append(g)
+	})
+
+	if err := h.Start(context.Background()); err != nil {
+		t.Fatal(err)
+	}
+
+	<-stop
+
+	if err := h.Stop(context.Background()); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// This test asserts that the context given to a one shot job cancels when the lifecycle of the group ends.
+func TestOneShot_LongRun(t *testing.T) {
+	started := make(chan struct{})
+	stopped := make(chan struct{})
+
+	h := fixture(func(r Registry, l hive.Lifecycle) {
+		g := r.NewGroup()
+
+		g.Add(
+			OneShot("long", func(ctx context.Context) error {
+				close(started)
+				<-ctx.Done()
+				defer close(stopped)
+				return nil
+			}),
+		)
+
+		l.Append(g)
+	})
+
+	if err := h.Start(context.Background()); err != nil {
+		t.Fatal(err)
+	}
+
+	<-started
+
+	if err := h.Stop(context.Background()); err != nil {
+		t.Fatal(err)
+	}
+
+	<-stopped
+}
+
+// This test asserts that we will stop retrying after the retry limit
+func TestOneShot_RetryFail(t *testing.T) {
+	var (
+		g Group
+		i int
+	)
+
+	const retries = 3
+
+	h := fixture(func(r Registry, l hive.Lifecycle) {
+		g = r.NewGroup()
+
+		g.Add(
+			OneShot("retry-fail", func(ctx context.Context) error {
+				defer func() { i++ }()
+				return errors.New("Always error")
+			}, WithRetry(retries, workqueue.DefaultControllerRateLimiter())),
+		)
+
+		l.Append(g)
+	})
+
+	if err := h.Start(context.Background()); err != nil {
+		t.Fatal(err)
+	}
+
+	// Continue as soon as all jobs stopped
+	g.(*group).wg.Wait()
+
+	if err := h.Stop(context.Background()); err != nil {
+		t.Fatal(err)
+	}
+
+	// 1 for the initial run, and 3 retries
+	if i != retries+1 {
+		t.Fatalf("Retries = %d, Ran = %d", retries, i)
+	}
+}
+
+// This test asserts that the one shot jobs have a delay equal to the expected behavior of the passed in ratelimiter.
+func TestOneShot_RetryBackoff(t *testing.T) {
+	var (
+		g     Group
+		i     int
+		times []time.Time
+	)
+
+	const retries = 6
+
+	h := fixture(func(r Registry, l hive.Lifecycle) {
+		g = r.NewGroup()
+
+		g.Add(
+			OneShot("retry-backoff", func(ctx context.Context) error {
+				defer func() { i++ }()
+				times = append(times, time.Now())
+				return errors.New("Always error")
+			}, WithRetry(retries, workqueue.NewItemExponentialFailureRateLimiter(50*time.Millisecond, 10*time.Second))),
+		)
+
+		l.Append(g)
+	})
+
+	if err := h.Start(context.Background()); err != nil {
+		t.Fatal(err)
+	}
+
+	// Continue as soon as all jobs stopped
+	g.(*group).wg.Wait()
+
+	if err := h.Stop(context.Background()); err != nil {
+		t.Fatal(err)
+	}
+
+	var last time.Duration
+	for i := 1; i < len(times); i++ {
+		diff := times[i].Sub(times[i-1])
+		if i > 2 {
+			// Test that the rate of change is 2 +- 50%, the 50% to account for CI time dilation.
```

Comment:
A bit torn here.I've had to remove checks like this in the past as I couldn't make them reliable enough, even with plenty of headroom. I'm pretty sure this will end up being flaky as VMs often have huge time dilation.

How about writing all the tests here that deal with time to retry multiple times on failures? This way a temporary "time dilation" won't screw us over.